compiler/rustc_middle/src/ty/subst.rs

```diff
@@ -103,6 +103,30 @@ impl<'tcx> GenericArgKind<'tcx> {
 
         GenericArg { ptr: unsafe { NonZeroUsize::new_unchecked(ptr | tag) }, marker: PhantomData }
     }
+
+    #[inline]
+    pub fn as_type(self) -> Option<Ty<'tcx>> {
+        match self {
+            GenericArgKind::Type(ty) => Some(ty),
+            _ => None,
+        }
+    }
+
+    #[inline]
+    pub fn as_region(self) -> Option<ty::Region<'tcx>> {
+        match self {
+            GenericArgKind::Lifetime(re) => Some(re),
+            _ => None,
+        }
+    }
+
+    #[inline]
+    pub fn as_const(self) -> Option<ty::Const<'tcx>> {
+        match self {
+            GenericArgKind::Const(ct) => Some(ct),
+            _ => None,
+        }
+    }
```

Comment:
Should those methods be on `GenericArgKind`, or directly on `GenericArg`? Seems like we always unpack first.

Reply:
I implemented them on the kind because it technically supports more use cases, but given that we always unpack first, I can move it to the `GenericArg`...